File: src/services/blockchainService.ts
import algosdk from 'algosdk';

export interface Badge {
  id: string;
  name: string;
  description: string;
  imageUrl: string;
  criteria: string;
  rarity: 'common' | 'uncommon' | 'rare' | 'epic' | 'legendary';
  assetId?: number;
  transactionId?: string;
}

export interface NFTCollectible {
  id: string;
  name: string;
  description: string;
  imageUrl: string;
  animationUrl?: string;
  storyMoment: string;
  rarity: 'common' | 'uncommon' | 'rare' | 'epic' | 'legendary';
  assetId?: number;
  transactionId?: string;
}

class BlockchainService {
  private algodClient: algosdk.Algodv2;
  private indexerClient: algosdk.Indexer;
  private creatorAccount: algosdk.Account | null = null;

  constructor() {
    // Initialize Algorand clients (using Nodely endpoints)
    const algodToken = import.meta.env.VITE_ALGORAND_TOKEN || '';
    const algodServer = import.meta.env.VITE_ALGORAND_SERVER || 'https://testnet-api.algonode.cloud';
    const algodPort = '';

    const indexerToken = '';
    const indexerServer = import.meta.env.VITE_ALGORAND_INDEXER || 'https://testnet-idx.algonode.cloud';
    const indexerPort = '';

    this.algodClient = new algosdk.Algodv2(algodToken, algodServer, algodPort);
    this.indexerClient = new algosdk.Indexer(indexerToken, indexerServer, indexerPort);
  }

  async initializeCreatorAccount(): Promise<void> {
    try {
      // In production, load from secure storage
      const mnemonic = import.meta.env.VITE_CREATOR_MNEMONIC;
      if (mnemonic) {
        this.creatorAccount = algosdk.mnemonicToSecretKey(mnemonic);
      } else {
        // Generate new account for demo
        this.creatorAccount = algosdk.generateAccount();
        console.log('Generated new creator account:', this.creatorAccount.addr);
      }
    } catch (error) {
      console.error('Error initializing creator account:', error);
    }
  }

  async createBadgeNFT(badge: Badge, recipientAddress: string): Promise<Badge | null> {
    if (!this.creatorAccount) {
      await this.initializeCreatorAccount();
    }

    if (!this.creatorAccount) {
      console.error('Creator account not initialized');
      return null;
    }

    try {
      const params = await this.algodClient.getTransactionParams().do();
      
      const assetCreateTxn = algosdk.makeAssetCreateTxnWithSuggestedParamsFromObject({
        from: this.creatorAccount.addr,
        suggestedParams: params,
        defaultFrozen: false,
        unitName: 'BADGE',
        assetName: badge.name,
        manager: this.creatorAccount.addr,
        reserve: this.creatorAccount.addr,
        freeze: this.creatorAccount.addr,
        clawback: this.creatorAccount.addr,
        assetURL: badge.imageUrl,
        assetMetadataHash: undefined,
        total: 1, // NFT - only one copy
        decimals: 0,
      });

      const signedTxn = assetCreateTxn.signTxn(this.creatorAccount.sk);
      const { txId } = await this.algodClient.sendRawTransaction(signedTxn).do();
      
      // Wait for confirmation
      await this.waitForConfirmation(txId);
      
      // Get asset ID
      const ptx = await this.algodClient.pendingTransactionInformation(txId).do();
      const assetId = ptx['asset-index'];

      // Transfer to recipient
      await this.transferAsset(assetId, recipientAddress);

      return {
        ...badge,
        assetId,
        transactionId: txId
      };
    } catch (error) {
      console.error('Error creating badge NFT:', error);
      return null;
    }
  }

  async createStoryNFT(collectible: NFTCollectible, recipientAddress: string): Promise<NFTCollectible | null> {
    if (!this.creatorAccount) {
      await this.initializeCreatorAccount();
    }

    if (!this.creatorAccount) {
      console.error('Creator account not initialized');
      return null;
    }

    try {
      const params = await this.algodClient.getTransactionParams().do();
      
      const metadata = {
        name: collectible.name,
        description: collectible.description,
        image: collectible.imageUrl,
        animation_url: collectible.animationUrl,
        attributes: [
          { trait_type: 'Rarity', value: collectible.rarity },
          { trait_type: 'Story Moment', value: collectible.storyMoment },
          { trait_type: 'Type', value: 'Story Collectible' }
        ]
      };

      const assetCreateTxn = algosdk.makeAssetCreateTxnWithSuggestedParamsFromObject({
        from: this.creatorAccount.addr,
        suggestedParams: params,
        defaultFrozen: false,
        unitName: 'STORY',
        assetName: collectible.name,
        manager: this.creatorAccount.addr,
        reserve: this.creatorAccount.addr,
        freeze: this.creatorAccount.addr,
        clawback: this.creatorAccount.addr,
        assetURL: collectible.imageUrl,
        assetMetadataHash: undefined,
        total: 1,
        decimals: 0,
      });

      const signedTxn = assetCreateTxn.signTxn(this.creatorAccount.sk);
      const { txId } = await this.algodClient.sendRawTransaction(signedTxn).do();
      
      await this.waitForConfirmation(txId);
      
      const ptx = await this.algodClient.pendingTransactionInformation(txId).do();
      const assetId = ptx['asset-index'];

      await this.transferAsset(assetId, recipientAddress);

      return {
        ...collectible,
        assetId,
        transactionId: txId
      };
    } catch (error) {
      console.error('Error creating story NFT:', error);
      return null;
    }
  }

  private async transferAsset(assetId: number, recipientAddress: string): Promise<void> {
    if (!this.creatorAccount) return;

    try {
      const params = await this.algodClient.getTransactionParams().do();
      
      // Opt-in transaction for recipient (in real app, recipient would do this)
      const optInTxn = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
        from: recipientAddress,
        to: recipientAddress,
        amount: 0,
        assetIndex: assetId,
        suggestedParams: params,
      });

      // Transfer transaction
      const transferTxn = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
        from: this.creatorAccount.addr,
        to: recipientAddress,
        amount: 1,
        assetIndex: assetId,
        suggestedParams: params,
      });

      const signedTransferTxn = transferTxn.signTxn(this.creatorAccount.sk);
      const { txId } = await this.algodClient.sendRawTransaction(signedTransferTxn).do();
      
      await this.waitForConfirmation(txId);
    } catch (error) {
      console.error('Error transferring asset:', error);
    }
  }

  private async waitForConfirmation(txId: string): Promise<void> {
    let response = await this.algodClient.status().do();
    let lastround = response['last-round'];
    
    while (true) {
      const pendingInfo = await this.algodClient.pendingTransactionInformation(txId).do();
      if (pendingInfo['confirmed-round'] !== null && pendingInfo['confirmed-round'] > 0) {
        break;
      }
      lastround++;
      await this.algodClient.statusAfterBlock(lastround).do();
    }
  }

  async getUserAssets(userAddress: string): Promise<(Badge | NFTCollectible)[]> {
    try {
      const accountInfo = await this.indexerClient.lookupAccountByID(userAddress).do();
      const assets = accountInfo.account.assets || [];
      
      // In a real implementation, you'd fetch metadata for each asset
      // For now, return mock data
      return [];
    } catch (error) {
      console.error('Error fetching user assets:', error);
      return [];
    }
  }

  // Demo function to simulate earning badges
  async simulateEarnBadge(badgeName: string, userAddress: string = 'demo'): Promise<Badge> {
    const badge: Badge = {
      id: `badge-${Date.now()}`,
      name: badgeName,
      description: `Earned for completing ${badgeName}`,
      imageUrl: `https://images.pexels.com/photos/1029141/pexels-photo-1029141.jpeg?auto=compress&cs=tinysrgb&w=200`,
      criteria: 'Complete the required learning objectives',
      rarity: 'uncommon',
      assetId: Math.floor(Math.random() * 1000000),
      transactionId: `demo-tx-${Date.now()}`
    };

    // Simulate blockchain transaction delay
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    return badge;
  }

  // Demo function to simulate earning story NFTs
  async simulateEarnNFT(storyMoment: string, userAddress: string = 'demo'): Promise<NFTCollectible> {
    const nft: NFTCollectible = {
      id: `nft-${Date.now()}`,
      name: `${storyMoment} Moment`,
      description: `A unique collectible from your story journey`,
      imageUrl: `https://images.pexels.com/photos/1029621/pexels-photo-1029621.jpeg?auto=compress&cs=tinysrgb&w=200`,
      animationUrl: `https://example.com/animations/story-${Date.now()}.mp4`,
      storyMoment,
      rarity: 'rare',
      assetId: Math.floor(Math.random() * 1000000),
      transactionId: `demo-tx-${Date.now()}`
    };

    await new Promise(resolve => setTimeout(resolve, 2000));
    
    return nft;
  }
}

export const blockchainService = new BlockchainService();

File: backend/app/services/blockchain_service.py
import algosdk
from typing import Dict, Any, Optional
import asyncio
from app.core.config import settings


class BlockchainService:
    """Blockchain service for NFT creation using Algorand"""
    
    def __init__(self):
        self.algod_client = self._init_algod_client()
        self.indexer_client = self._init_indexer_client()
        self.creator_account = self._init_creator_account()
    
    def _init_algod_client(self):
        """Initialize Algorand client"""
        try:
            return algosdk.v2client.algod.AlgodClient(
                settings.ALGORAND_TOKEN or "",
                settings.ALGORAND_SERVER,
                headers={"User-Agent": "litink-backend"}
            )
        except Exception as e:
            print(f"Algorand client initialization error: {e}")
            return None
    
    def _init_indexer_client(self):
        """Initialize Algorand indexer client"""
        try:
            return algosdk.v2client.indexer.IndexerClient(
                "",
                settings.ALGORAND_INDEXER,
                headers={"User-Agent": "litink-backend"}
            )
        except Exception as e:
            print(f"Algorand indexer initialization error: {e}")
            return None
    
    def _init_creator_account(self):
        """Initialize creator account"""
        try:
            if settings.CREATOR_MNEMONIC:
                return algosdk.mnemonic.to_private_key(settings.CREATOR_MNEMONIC)
            else:
                # Generate new account for demo
                private_key, address = algosdk.account.generate_account()
                print(f"Generated new creator account: {address}")
                return private_key
        except Exception as e:
            print(f"Creator account initialization error: {e}")
            return None
    
    async def create_badge_nft(
        self,
        name: str,
        description: str,
        image_url: Optional[str],
        recipient_address: str
    ) -> Optional[Dict[str, Any]]:
        """Create badge NFT on Algorand"""
        if not self.algod_client or not self.creator_account:
            return await self._mock_create_badge(name, description)
        
        try:
            # Get suggested parameters
            params = self.algod_client.suggested_params()
            
            # Create asset
            txn = algosdk.future.transaction.AssetConfigTxn(
                sender=algosdk.account.address_from_private_key(self.creator_account),
                sp=params,
                total=1,  # NFT - only one copy
                default_frozen=False,
                unit_name="BADGE",
                asset_name=name,
                manager=algosdk.account.address_from_private_key(self.creator_account),
                reserve=algosdk.account.address_from_private_key(self.creator_account),
                freeze=algosdk.account.address_from_private_key(self.creator_account),
                clawback=algosdk.account.address_from_private_key(self.creator_account),
                url=image_url or "",
                decimals=0
            )
            
            # Sign transaction
            signed_txn = txn.sign(self.creator_account)
            
            # Submit transaction
            tx_id = self.algod_client.send_transaction(signed_txn)
            
            # Wait for confirmation
            await self._wait_for_confirmation(tx_id)
            
            # Get asset ID
            ptx = self.algod_client.pending_transaction_info(tx_id)
            asset_id = ptx["asset-index"]
            
            return {
                "asset_id": asset_id,
                "transaction_id": tx_id,
                "status": "success"
            }
            
        except Exception as e:
            print(f"Blockchain service error: {e}")
            return await self._mock_create_badge(name, description)
    
    async def create_story_nft(
        self,
        name: str,
        description: str,
        image_url: Optional[str],
        story_moment: str,
        recipient_address: str
    ) -> Optional[Dict[str, Any]]:
        """Create story NFT on Algorand"""
        if not self.algod_client or not self.creator_account:
            return await self._mock_create_story_nft(name, description)
        
        try:
            # Similar to badge creation but with story-specific metadata
            params = self.algod_client.suggested_params()
            
            txn = algosdk.future.transaction.AssetConfigTxn(
                sender=algosdk.account.address_from_private_key(self.creator_account),
                sp=params,
                total=1,
                default_frozen=False,
                unit_name="STORY",
                asset_name=name,
                manager=algosdk.account.address_from_private_key(self.creator_account),
                reserve=algosdk.account.address_from_private_key(self.creator_account),
                freeze=algosdk.account.address_from_private_key(self.creator_account),
                clawback=algosdk.account.address_from_private_key(self.creator_account),
                url=image_url or "",
                decimals=0
            )
            
            signed_txn = txn.sign(self.creator_account)
            tx_id = self.algod_client.send_transaction(signed_txn)
            
            await self._wait_for_confirmation(tx_id)
            
            ptx = self.algod_client.pending_transaction_info(tx_id)
            asset_id = ptx["asset-index"]
            
            return {
                "asset_id": asset_id,
                "transaction_id": tx_id,
                "status": "success"
            }
            
        except Exception as e:
            print(f"Blockchain service error: {e}")
            return await self._mock_create_story_nft(name, description)
    
    async def _wait_for_confirmation(self, tx_id: str):
        """Wait for transaction confirmation"""
        try:
            last_round = self.algod_client.status()["last-round"]
            
            while True:
                pending_info = self.algod_client.pending_transaction_info(tx_id)
                if pending_info.get("confirmed-round", 0) > 0:
                    break
                
                last_round += 1
                self.algod_client.status_after_block(last_round)
                
        except Exception as e:
            print(f"Confirmation wait error: {e}")
    
    async def _mock_create_badge(self, name: str, description: str) -> Dict[str, Any]:
        """Mock badge creation for development"""
        await asyncio.sleep(2)  # Simulate blockchain transaction time
        
        return {
            "asset_id": hash(name) % 1000000,
            "transaction_id": f"mock_tx_{hash(name)}",
            "status": "success"
        }
    
    async def _mock_create_story_nft(self, name: str, description: str) -> Dict[str, Any]:
        """Mock story NFT creation for development"""
        await asyncio.sleep(2)
        
        return {
            "asset_id": hash(name) % 1000000,
            "transaction_id": f"mock_story_tx_{hash(name)}",
            "status": "success"
        }